moveit_commander/src/moveit_commander/planning_scene_interface.py

```diff
@@ -63,98 +62,32 @@ def __init__(self):
-        co = CollisionObject()
-        co.operation = CollisionObject.ADD
-        co.id = name
-        co.header = pose.header
-        box = SolidPrimitive()
-        box.type = SolidPrimitive.BOX
-        box.dimensions = list(size)
-        co.primitives = [box]
-        co.primitive_poses = [pose.pose]
-        return co
-    
-    def __make_mesh(self, name, pose, filename, scale = (1, 1, 1)):
-        co = CollisionObject()
-        if pyassimp is False:
-            raise MoveItCommanderException("Pyassimp needs patch https://launchpadlibrarian.net/319496602/patchPyassim.txt")
-        scene = pyassimp.load(filename)
-        if not scene.meshes or len(scene.meshes) == 0:
-            raise MoveItCommanderException("There are no meshes in the file")
-        if len(scene.meshes[0].faces) == 0:
-            raise MoveItCommanderException("There are no faces in the mesh")
-        co.operation = CollisionObject.ADD
-        co.id = name
-        co.header = pose.header
-        
-        mesh = Mesh()
-        first_face = scene.meshes[0].faces[0]
-        if hasattr(first_face, '__len__'):
-            for face in scene.meshes[0].faces:
-                if len(face) == 3:
-                    triangle = MeshTriangle()
-                    triangle.vertex_indices = [face[0], face[1], face[2]]
-                    mesh.triangles.append(triangle)
-        elif hasattr(first_face, 'indices'):
-            for face in scene.meshes[0].faces:
-                if len(face.indices) == 3:
-                    triangle = MeshTriangle()
-                    triangle.vertex_indices = [face.indices[0],
-                                               face.indices[1],
-                                               face.indices[2]]
-                    mesh.triangles.append(triangle)
-        else:
-            raise MoveItCommanderException("Unable to build triangles from mesh due to mesh object structure")
-        for vertex in scene.meshes[0].vertices:
-            point = Point()
-            point.x = vertex[0]*scale[0]
-            point.y = vertex[1]*scale[1]
-            point.z = vertex[2]*scale[2]
-            mesh.vertices.append(point)
-        co.meshes = [mesh]
-        co.mesh_poses = [pose.pose]
-        pyassimp.release(scene)
-        return co
-    
-    def __make_existing(self, name):
-        """
-        Create an empty Collision Object, used when the object already exists 
-        """
-        co = CollisionObject()
-        co.id = name
-        return co
-
     def add_mesh(self, name, pose, filename, size = (1, 1, 1)):
         """
         Add a mesh to the planning scene
         """
-        self._pub_co.publish(self.__make_mesh(name, pose, filename, size))
+        collision_object = self.__make_mesh(name, pose, filename, size)
+        diff_req = self.__make_planning_scene_diff_req(collision_object)
+        self._apply_planning_scene_diff.call(diff_req)
+        self._pub_co.publish(collision_object)
```

Comment:
This is redundant and I do not believe it is a good idea to publish *and* apply.
Actually, in my opinion this should even trigger warnings saying "adding object which already exists", but I think we do not do that to make the `move_group` more robust.
I created `apply_planning_scene` as a synchronous alternative to the corresponding topic interfaces, not as an overall replacement.
The shortcoming of the service is that it does not support to update multiple `PlanningSceneMonitor`s simultaneously and I believe some users require this.

In the C++ API, I implemented this via `apply*` functions in contrast to the topic-based `add*` updates.
For the Python API, I would propose to add an optional parameter `synchronous`.
Would this work for you?

Reply:
You are right, we should not do both at once. This was a change I've made to make sure my setup works when starting everything from the same launch file.

When I add a ton of sleep the topic interface works too. 

Btw looking around the code here I've discovered that `PlanningSceneInterface::addCollisionObjects` is not called at all within MoveIt. The `PlanningSceneMonitor` uses thet methods of `PlanningScene` directly.
```
$ grep -R addCollisionObjects
moveit_ros/planning_interface/planning_scene_interface/include/moveit/planning_scene_interface/planning_scene_interface.h:  void addCollisionObjects(
moveit_ros/planning_interface/planning_scene_interface/src/planning_scene_interface.cpp:  void addCollisionObjects(const std::vector<moveit_msgs::CollisionObject>& collision_objects,
moveit_ros/planning_interface/planning_scene_interface/src/planning_scene_interface.cpp:void PlanningSceneInterface::addCollisionObjects(const std::vector<moveit_msgs::CollisionObject>& collision_objects,
moveit_ros/planning_interface/planning_scene_interface/src/planning_scene_interface.cpp:  impl_->addCollisionObjects(collision_objects, object_colors);
```

Reply:
Added 2 parameters to __init__:
 * `asynchronous=False`
 * `service_timeout=5.0`
and a `__submit()` function that decides where to let the message out depending on these params :)